client/src/screens/auth/SignUp.js

```diff
@@ -0,0 +1,152 @@
+import React, { Component } from 'react';
+import { Keyboard } from 'react-native';
+import { connect } from 'react-redux';
+import { graphql, compose } from 'react-apollo';
+import PropTypes from 'prop-types';
+
+import { setCurrentUser } from '../../state/auth.actions';
+
+import { Container, Holder } from '../../components/Container';
+import { Header } from '../../components/Header';
+import { Input } from '../../components/TextInput';
+import { NavBar } from '../../components/NavBar';
+import { NextButton } from '../../components/Button';
+import { Alert } from '../../components/Alert';
+
+import SIGNUP_MUTATION from '../../graphql/signup.mutation';
+
+class SignUp extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      username: '',
+      password: '',
+      email: '',
+      usernameValid: false,
+      passwordValid: false,
+      emailValid: false,
+    };
+    this.onPressSignUp = this.onPressSignUp.bind(this);
+  }
+
+  onPressSignUp() {
+    const { deviceUuid } = this.props.local;
+    const { username, password, email } = this.state;
+    this.props
+      .signup({ username, email, password, deviceUuid })
+      .then(({ data: { signup: user } }) => {
+        const ourUser = {
+          username: user.username,
+          token: user.authToken,
+        };
+        this.props.dispatch(setCurrentUser(ourUser));
+      })
+      .catch((error) => {
+        this.setState({
+          status: 'Sign Up Error',
+          errorMessage: error.message,
+        });
+        Keyboard.dismiss();
+        this.popRef.show();
+      });
+  }
+
+  onChangeEmail = (value) => {
+    let emailValid = false;
+    if (value.indexOf('@') !== -1) emailValid = true;
+
+    this.setState({
+      email: value,
+      emailValid,
+    });
+  };
+
+  onChangePassword = (value) => {
+    let passwordValid = false;
+    if (value.length > 5) passwordValid = true;
+
+    this.setState({
+      password: value,
+      passwordValid,
+    });
+  };
+
+  onChangeUsername = (value) => {
+    let usernameValid = false;
+    if (value.length > 5) usernameValid = true;
+
+    this.setState({
+      username: value,
+      usernameValid,
+    });
+  };
+
+  render() {
+    return (
+      <Container>
+        <NavBar onPressBack={() => this.props.navigation.goBack()} />
```

Comment:
use bind in constructor instead of callbacks to avoid constructing extra functions. Every time `render()` is called it will re-create these anonymous functions.

Reply:
Moved to function with new ES7 syntax to avoid binding call in constructor.